Use MathML attributes for PDFs read in Adobe Acrobat

NSoiffer opened the pull request:
### Link to issue number:
This fixes #17980

### Summary of the issue:
For PDF, the code was not grabbing MathML attributes needed for speech.

### Description of user facing changes
The speech for math, and rarely the braille (e.g., bevelled fractions in Nemeth), was not always correct due to attributes not being picked up and the defaults being used.

### Description of development approach
Unfortunately, the PDF interface does not allow grabbing all the attributes. Instead, one must ask for each attribute individually. Most attributes don't affect speech or braille, so it is not necessary to get them. I looked at what MathCAT used and added those. In particular, the following are picked up:
```
		id = node.GetID()
		if id:
			yield f' id="{id}"'
		yield getMathMLAttributes(node, ["intent", "arg"])
		if tag == "mi" or tag == "mn" or tag == "mo" or tag == "mtext":
			yield getMathMLAttributes(node, ["mathvariant"])
		elif tag == "mfenced":
			yield getMathMLAttributes(node, ["open", "close", "separators"])
		elif tag == "menclose":
			yield getMathMLAttributes(node, ["notation"])
		elif tag == "annotation-xml" or tag == "annotation":
			yield getMathMLAttributes(node, ["encoding"])
		elif tag == "ms":
			yield getMathMLAttributes(node, ["open", "close"])
```
Note: `intent` and `arg` are new to MathML 4 and are aimed at improving speech for math.

### Testing strategy:
There is a test file in the issue. It tests `intent`. I asked David Carlisle, a LaTeX developer, to generate some (fully tagged) PDF examples that use more than `intent` (which is what pdftex will generate on its own in some cases. He gave me a sample that displays poorly, but has some attrs hacked into it. There is a log statement (at debug level) that shows the MathML that is gathered up. The MathML picked was correct. The speech was also as expected given the MathML.

I don't know how one would create a unit or system test for this. The issue contains a PDF example which can be tested (issue describes the expected result).

### Known issues with pull request:
None.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

<!-- Please keep the following -->
@coderabbitai summary

codeofdusk commented:
@NSoiffer Should this target `beta` (i.e. is it a bug in #17276)?

NSoiffer commented:
It is a bug that results in incorrect speech and possibly incorrect braille if one of the attributes for the elements listed above is present. For example, if someone uses
```
<mi mathvariant="double-struck">C</mi>
```
then instead of hearing either "the complex numbers" or "double struck cap C", they will just hear "cap C". Similarly, the braille will be wrong as it will be missing a script indicator. Or as in the bug report, they won't hear the equation label as a label ("line 2 with label 2 ...") and I think they will hear the label as just data which is very confusing.

My feeling this is a pretty serious bug and should be part of the beta.

codeofdusk commented:
In which case, please change the base branch on GitHub and rebase your branch accordingly.

seanbudd commented on source/NVDAObjects/IAccessible/adobeAcrobat.py:
```diff
@@ -144,9 +166,28 @@ def _getNodeMathMl(self, node):
+			try:
+				child = self.pdDomNode.GetChild(childNum).QueryInterface(IPDDomElement)
+			except COMError:
+				log.debugWarning(f"COMError trying to get childNum={childNum}")
```
```suggestion
				log.debugWarning(f"COMError trying to get {childNum=}")
```

NSoiffer replied:
I don't think your suggestion makes sense. You want an error message like "COMError trying to get childNum=1".

seanbudd commented on source/NVDAObjects/IAccessible/adobeAcrobat.py:
```diff
@@ -144,9 +166,28 @@ def _getNodeMathMl(self, node):
+				log.debug(f"\t(PDF) get_mathMl: tag={child.GetTagName()}")
+			if child.GetTagName() == "math":
+				answer = "".join(self._getNodeMathMl(child))
+				if log.isEnabledFor(log.DEBUG):
+					log.debug(f"_get_mathMl: found MathML = {answer}")
```
Nothing is being calculated here, so there is no need to check `isEnabledFor`

```suggestion
				log.debug(f"_get_mathMl: found MathML = {answer}")
```

NSoiffer replied:
Fixed

NSoiffer commented on source/NVDAObjects/IAccessible/adobeAcrobat.py:
```diff
@@ -120,14 +120,36 @@ def _isEqual(self, other):
+		if id:
+			yield f' id="{id}"'
+		yield getMathMLAttributes(node, ["intent", "arg"])
+		if tag == "mi" or tag == "mn" or tag == "mo" or tag == "mtext":
+			yield getMathMLAttributes(node, ["mathvariant"])
+		elif tag == "mfenced":
+			yield getMathMLAttributes(node, ["open", "close", "separators"])
+		elif tag == "menclose":
+			yield getMathMLAttributes(node, ["notation"])
+		elif tag == "annotation-xml" or tag == "annotation":
+			yield getMathMLAttributes(node, ["encoding"])
+		elif tag == "ms":
+			yield getMathMLAttributes(node, ["open", "close"])
```
Sorry, I should have used match/case. It just grew organically and I should have switched at some point.

NSoiffer commented on source/NVDAObjects/IAccessible/adobeAcrobat.py:
```diff
@@ -120,14 +120,36 @@ def _isEqual(self, other):
 	def _getNodeMathMl(self, node):
+		"""Traverse the MathML tree and return an XML string representing the math"""
+
+		def getMathMLAttributes(node, attrList: list) -> str:
```
Since it is only used in `_getNodeMathMl`, I'm a fan of limiting its scope. I changed the parameter `node` so there is no confusion with shadowing.

seanbudd replied:
the zen of python says flat is better than nested. Using nested functions increases complexity. Can you please move it out?

NSoiffer replied:
I've moved it out. Personally, I think it is uglier (now calls a static method of the class), but I'm not a native python speaker.

FYI: I leave for a holiday in 14 hours. Please feel free to correct any remaining issues. AFAIK, all of the comments are resolved.

NSoiffer commented on source/NVDAObjects/IAccessible/adobeAcrobat.py:
```diff
@@ -120,14 +120,36 @@ def _isEqual(self, other):
+
+		def getMathMLAttributes(node, attrList: list) -> str:
+			attrValues = ""
+			# log.info(f"\nstart getMathMLAttributes -- looking for: {attrList}")
```
I guess it is my personal preference: if I found some debugging statement useful when writing the code, I tend to leave it in because it might be useful in the future. However, I do understand it can make the code more cluttered, so I've removed most of them and just left in the ones as `log.debug` that show what the returned MathML is and where it came from.

seanbudd replied:
Can you please remove this, or uncomment it

NSoiffer commented on source/NVDAObjects/IAccessible/adobeAcrobat.py:
```diff
@@ -120,14 +120,36 @@ def _isEqual(self, other):
 		return super(AcrobatNode, self)._isEqual(other)
 
 	def _getNodeMathMl(self, node):
```
I looked at the type of 'node' and it is some very messy (long) type. For the output, I don't know the type hint for a generator function (still unsure after looking it up). However, the strings really aren't that long (very likely less than 200 chars, and almost certainly less than 1000 chars), so I converted it to return a string. Let me know if you don't feel that is appropriate.

NSoiffer committed:
Rebased file on beta branch.

Updated adobeAcrobat.py with suggestions as per the PR
Updated changes.md as per the PR

Fingers crossed I got this right...

NSoiffer commented:
I _think_ I have this rebased to the beta and made all the changes suggested. This is at the edge of my github knowledge, so hopefully it is ok.

NSoiffer committed:
Moved `getMathMLAttributes` out from being a nested function to being a static class method.

Added a few more comments.

seanbudd commented:
This has been retargeted to beta, but not rebased off beta. It includes commits from master which cannot be merged. You need to drop all the commits from master in the rebase

SaschaCowley commented:
@NSoiffer do you need help correctly rebasing this?

SaschaCowley committed:
Update source/NVDAObjects/IAccessible/adobeAcrobat.py

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

seanbudd commented on source/NVDAObjects/IAccessible/adobeAcrobat.py:
```diff
@@ -119,61 +119,96 @@ def _isEqual(self, other):
 
-	def _getNodeMathMl(self, node):
+	@staticmethod
+	def getMathMLAttributes(node, attrList: list) -> str:
```
Node still needs a type hint

```suggestion
	def getMathMLAttributes(node: IPDDomElement, attrList: list) -> str:
```

seanbudd commented on source/NVDAObjects/IAccessible/adobeAcrobat.py:
```diff
@@ -119,61 +119,96 @@ def _isEqual(self, other):
+				attrValues += f' {attr}="{val}"'
+		return attrValues
+
+	def _getNodeMathMl(self, node) -> str:
```
```suggestion
	def _getNodeMathMl(self, node: IPDDomElement) -> str:
```

seanbudd commented on source/NVDAObjects/IAccessible/adobeAcrobat.py:
```diff
@@ -119,61 +119,96 @@ def _isEqual(self, other):
+					answer += sub
+		return answer + f"</{tag}>"
 
 	def _get_mathMl(self) -> str:
```
this seems like it can return None, should it return an empty string at the end or should the type hint be updated?

davidcarlisle replied:
if it doesn't raise an exception, `_get_mathML` always returns a `str` doesn't it (unless I am misreading the code the three possible `return` are all strings?

seanbudd replied:
Nevermind - yes I misread this

seanbudd committed:
Apply suggestions from code review

davidcarlisle commented:
@SaschaCowley  @seanbudd  thanks for picking this up in Neil's absence. With 2025.1  we will be able to make mathematical PDF that may be read in acrobat and foxit with the mathematical content read to the same level as web pages, this PR is a last missing piece for acrobat.  I'm so happy to see it merged.
